app/services/experimentation/data_export.rb

```diff
@@ -0,0 +1,195 @@
+
+    def initialize(name)
+      @experiment = Experimentation::Experiment.find_by!(name: name)
+      remind_ons = experiment.reminder_templates.pluck(:remind_on_in_days)
+      @notification_start_date = experiment.start_date - remind_ons.min.days
+      @max_communications = 0
+      @max_appointments = 0
+      @max_past_visits = 0
+      @aggregate = []
+      @cutoff_date = experiment.end_date + FOLLOWUP_CUTOFF
+    end
+
+    def as_csv
+      aggregate_data
+      pad_communications
+      pad_appointments
+      pad_past_visits
+      create_csv
+    end
+
+    private
+
+    def aggregate_data
+      experiment.treatment_groups.each do |group|
+        group.patients.each do |patient|
+          tgm = patient.treatment_group_memberships.find_by(treatment_group_id: group.id)
+          notifications = patient.notifications.where(experiment_id: experiment.id).order(:remind_on)
+          assigned_facility = patient.assigned_facility
+
+          aggregate << {
+            experiment_name: experiment.name,
+            treatment_group: group.description,
+            experiment_inclusion_date: tgm.created_at.to_date,
+            appointments: process_appointments(patient, notifications),
+            encounters: process_past_visits(patient),
+            communications: process_communications(notifications),
+            patient_gender: patient.gender,
+            patient_age: patient.age,
+            patient_risk_level: patient.risk_priority,
+            diagnosed_hypertensive: true, # remove?
+            patient_has_phone: true, # remove?
+            assigned_facility_name: assigned_facility&.name,
+            assigned_facility_type: assigned_facility&.facility_type,
+            assigned_facility_state: assigned_facility&.state,
+            assigned_facility_district: assigned_facility&.district,
+            assigned_facility_block: assigned_facility&.block,
+            patient_registration_date: patient.device_created_at.to_date,
+            patient_id: tgm.id
+          }
+        end
+      end
+    end
+
+    def process_communications(notifications)
+      communications = notifications.each_with_object([]) do |notification, communications_data|
+        ordered_communications = notification.communications.sort_by{|c| c.created_at }
+        ordered_communications.each do |c|
+          communications_data << [notification.message, c.communication_type, c.detailable&.delivered_on, c.detailable&.result]
+        end
+      end
+
+      @max_communications = communications.count if communications.count > @max_communications
+      communications
+    end
+
+    def pad_communications
+      aggregate.each do |patient_data|
+        communications_deficit = @max_communications - patient_data[:communications].count
+        communications_deficit.times do
+          patient_data[:communications] << Array.new(4, nil)
+        end
+      end
+    end
+
+    def process_appointments(patient, notifications)
```

Comment:
This is the most complex and ambiguous portion of the export because we don't have a clear connection between appointment and encounter. For a first pass, I've just selected all encounters from the first day of notifications til ten days after experiment completion. This should probably be refined to only look for encounters within a certain date range of the appointment. Per my comment above, I think we discussed a ten day cutoff after the appointment, but I don't recall any discussion of a cutoff before the appointment.